src/mscorlib/shared/System/IO/PathHelper.Windows.cs

```diff
@@ -89,173 +76,152 @@ private static int GetInputBuffer(ref StringBuffer content, bool isDosUnc, ref S
-                            outputBuffer.Append(ref inputBuffer, foundIndex, inputBuffer.Length - foundIndex);
-                        }
+                        // Temporarily set a null in the string to get Windows to look further up the path
+                        inputBuilder[foundIndex] = '\0';
                     }
                 }
-
-                // Strip out the prefix and return the string
-                ref StringBuffer bufferToUse = ref Choose(success, ref outputBuffer, ref inputBuffer);
-
-                // Switch back from \\?\ to \\.\ if necessary
-                if (wasDotDevice)
-                    bufferToUse[2] = '.';
-
-                string returnValue = null;
-
-                int newLength = bufferToUse.Length - rootDifference;
-                if (isDosUnc)
-                {
-                    // Need to go from \\?\UNC\ to \\?\UN\\
-                    bufferToUse[PathInternal.UncExtendedPrefixLength - PathInternal.UncPrefixLength] = '\\';
-                }
-
-                // We now need to strip out any added characters at the front of the string
-                if (bufferToUse.SubstringEquals(originalPath, rootDifference, newLength))
+                else if (result > outputBuilder.Capacity)
                 {
-                    // Use the original path to avoid allocating
-                    returnValue = originalPath;
+                    // Not enough space. The result count for this API does not include the null terminator.
+                    outputBuilder.EnsureCapacity(checked((int)result));
+                    result = Interop.Kernel32.GetLongPathNameW(ref inputBuilder.GetPinnableReference(), ref outputBuilder.GetPinnableReference(), (uint)outputBuilder.Capacity);
                 }
                 else
                 {
-                    returnValue = bufferToUse.Substring(rootDifference, newLength);
+                    // Found the path
+                    success = true;
+                    outputBuilder.Length = checked((int)result);
+                    if (foundIndex < inputLength - 1)
+                    {
+                        // It was a partial find, put the non-existent part of the path back
+                        outputBuilder.Append(inputBuilder.AsSpan().Slice(foundIndex, inputBuilder.Length - foundIndex));
+                    }
                 }
-
-                return returnValue;
-            }
-            finally
-            {
-                inputBuffer.Free();
             }
-        }
 
-        // Helper method to workaround lack of operator ? support for ref values
-        private static ref StringBuffer Choose(bool condition, ref StringBuffer s1, ref StringBuffer s2)
-        {
-            if (condition) return ref s1;
-            else return ref s2;
+            // If we were able to expand the path, use it, otherwise use the original full path result
+            outputBuilder = success ? ref outputBuilder : ref inputBuilder;
```

Comment:
This will *overwrite* the outputBuilder with inputBuilder. The result is going to be that we are not going to release the buffer from the original outputBuilder if there is one.

I think you still need `ref StringBuffer bufferToUse` local here like before.

Reply:
D'oh.  If I take out the ref on the method sig will that solve it?

Reply:
On second thought, that might not be a good idea. I could see that getting reintroduced.

Reply:
I do not think it would solve it because of you are mutating the outputBuilder here.

Also, I do not think it is a generally good idea to play tricks when you do copy vs. when you do not make a copy of "no copy" types. It is best to always pass the "no copy" types by ref. I wish there was analyzer to check for copies of no-copy types.